File: bindings/lua/causalontology/json.lua
-- json.lua - a shape-preserving JSON layer for causalontology-lua.
--
-- Lua tables give string keys no order and cannot tell an empty array from
-- an empty object, and plain tonumber() would erase the integer-versus-
-- decimal source distinction the canonicalizer needs.  This module therefore
-- decodes JSON into:
--
--   * objects: tables tagged with a private ORDER key holding the insertion
--     order of their string keys (mirroring Python's insertion-ordered dict),
--   * arrays:  tables tagged with a private IS_ARRAY key,
--   * numbers: Lua 5.4 integers when the literal carries no '.', 'e' or 'E'
--     (via math.tointeger), Lua floats otherwise,
--   * null:    the json.null sentinel (Lua nil would delete the key).
--
-- Every other module iterates objects only through json.keys(), never
-- through pairs(), so behavior is deterministic everywhere store.py
-- iterates dicts (objects, records, cycle-finder nodes, view buckets).

local json = {}

-- Private, non-string sentinel keys: they can never collide with JSON keys.
json.ORDER = setmetatable({}, { __tostring = function() return "<order>" end })
json.IS_ARRAY = setmetatable({}, { __tostring = function() return "<array>" end })
json.null = setmetatable({}, { __tostring = function() return "null" end })

-- ---------------------------------------------------------------- objects

-- A new, empty, insertion-ordered object.
function json.new_object()
  local t = {}
  t[json.ORDER] = {}
  return t
end

function json.is_object(t)
  return type(t) == "table" and rawget(t, json.ORDER) ~= nil
end

-- Set key k to value v, appending k to the order on first appearance.
-- Setting v = nil removes the key (and its order slot).
function json.set(obj, k, v)
  local order = obj[json.ORDER]
  if v == nil then
    if obj[k] ~= nil then
      obj[k] = nil
      for i, name in ipairs(order) do
        if name == k then table.remove(order, i) break end
      end
    end
    return obj
  end
  if obj[k] == nil then order[#order + 1] = k end
  obj[k] = v
  return obj
end

-- Set key k only when it is not already present (Python's dict.setdefault).
function json.setdefault(obj, k, v)
  if obj[k] == nil then json.set(obj, k, v) end
  return obj[k]
end

-- The insertion-order key list (the live array; callers must not mutate it).
function json.keys(obj)
  return obj[json.ORDER]
end

-- A shallow copy preserving insertion order (Python's dict(record)).
function json.copy_object(obj)
  local out = json.new_object()
  for _, k in ipairs(obj[json.ORDER]) do
    json.set(out, k, obj[k])
  end
  return out
end

-- Build an ordered object from a flat list of key, value pairs.
function json.obj(...)
  local out = json.new_object()
  local args = table.pack(...)
  for i = 1, args.n, 2 do
    json.set(out, args[i], args[i + 1])
  end
  return out
end

-- ----------------------------------------------------------------- arrays

-- A new tagged array, optionally wrapping an existing sequence.
function json.new_array(list)
  local t = list or {}
  t[json.IS_ARRAY] = true
  return t
end

function json.is_array(t)
  return type(t) == "table" and rawget(t, json.IS_ARRAY) ~= nil
end

-- A shallow copy of a tagged array.
function json.copy_array(arr)
  local out = json.new_array()
  for i, v in ipairs(arr) do out[i] = v end
  return out
end

-- ------------------------------------------------------------ deep equal

-- Deep structural equality with Python semantics: objects compare by key
-- set and values (order-insensitive), arrays elementwise, and 1 == 1.0.
function json.deep_equal(a, b)
  if a == b then return true end
  if json.is_object(a) and json.is_object(b) then
    local ka, kb = json.keys(a), json.keys(b)
    if #ka ~= #kb then return false end
    for _, k in ipairs(ka) do
      if b[k] == nil or not json.deep_equal(a[k], b[k]) then return false end
    end
    return true
  end
  if json.is_array(a) and json.is_array(b) then
    if #a ~= #b then return false end
    for i = 1, #a do
      if not json.deep_equal(a[i], b[i]) then return false end
    end
    return true
  end
  return false
end

-- ----------------------------------------------------------------- parser

local function parse_error(text, pos, why)
  error(string.format("JSON parse error at byte %d: %s", pos, why), 0)
end

local function skip_ws(text, pos)
  local _, stop = text:find("^[ \t\r\n]*", pos)
  return stop + 1
end

-- Encode one Unicode code point as UTF-8 bytes (for \uXXXX escapes).
local function utf8_encode(cp)
  return utf8.char(cp)
end

local function parse_string(text, pos)
  -- pos points at the opening quote
  local out, i = {}, pos + 1
  while true do
    local c = text:sub(i, i)
    if c == "" then parse_error(text, i, "unterminated string") end
    if c == '"' then
      return table.concat(out), i + 1
    elseif c == "\\" then
      local esc = text:sub(i + 1, i + 1)
      if esc == '"' then out[#out + 1] = '"'; i = i + 2
      elseif esc == "\\" then out[#out + 1] = "\\"; i = i + 2
      elseif esc == "/" then out[#out + 1] = "/"; i = i + 2
      elseif esc == "b" then out[#out + 1] = "\b"; i = i + 2
      elseif esc == "f" then out[#out + 1] = "\f"; i = i + 2
      elseif esc == "n" then out[#out + 1] = "\n"; i = i + 2
      elseif esc == "r" then out[#out + 1] = "\r"; i = i + 2
      elseif esc == "t" then out[#out + 1] = "\t"; i = i + 2
      elseif esc == "u" then
        local hex = text:sub(i + 2, i + 5)
        if not hex:match("^%x%x%x%x$") then
          parse_error(text, i, "bad \\u escape")
        end
        local cp = tonumber(hex, 16)
        i = i + 6
        -- combine a UTF-16 surrogate pair into one code point
        if cp >= 0xD800 and cp <= 0xDBFF and text:sub(i, i + 1) == "\\u" then
          local hex2 = text:sub(i + 2, i + 5)
          local lo = tonumber(hex2, 16)
          if lo and lo >= 0xDC00 and lo <= 0xDFFF then
            cp = 0x10000 + ((cp - 0xD800) << 10) + (lo - 0xDC00)
            i = i + 6
          end
        end
        out[#out + 1] = utf8_encode(cp)
      else
        parse_error(text, i, "bad escape")
      end
    else
      out[#out + 1] = c
      i = i + 1
    end
  end
end

local parse_value  -- forward declaration

local function parse_number(text, pos)
  -- capture the whole numeric literal so its shape decides the Lua subtype
  local lit = text:match("^-?%d+%.?%d*[eE]?[%+%-]?%d*", pos)
  -- trim any over-capture from the loose pattern above
  lit = lit:match("^-?%d+%.%d+[eE][%+%-]?%d+")
        or lit:match("^-?%d+%.%d+")
        or lit:match("^-?%d+[eE][%+%-]?%d+")
        or lit:match("^-?%d+")
  if not lit then parse_error(text, pos, "bad number") end
  local n = tonumber(lit)
  -- the shape rule: integer when the literal has no '.', 'e' or 'E'
  if not lit:find("[%.eE]") then
    n = math.tointeger(n) or n
  end
  return n, pos + #lit
end

local function parse_array(text, pos)
  local arr = json.new_array()
  pos = skip_ws(text, pos + 1)
  if text:sub(pos, pos) == "]" then return arr, pos + 1 end
  while true do
    local v
    v, pos = parse_value(text, pos)
    arr[#arr + 1] = v
    pos = skip_ws(text, pos)
    local c = text:sub(pos, pos)
    if c == "," then
      pos = skip_ws(text, pos + 1)
    elseif c == "]" then
      return arr, pos + 1
    else
      parse_error(text, pos, "expected ',' or ']'")
    end
  end
end

local function parse_object(text, pos)
  local obj = json.new_object()
  pos = skip_ws(text, pos + 1)
  if text:sub(pos, pos) == "}" then return obj, pos + 1 end
  while true do
    if text:sub(pos, pos) ~= '"' then
      parse_error(text, pos, "expected object key")
    end
    local k, v
    k, pos = parse_string(text, pos)
    pos = skip_ws(text, pos)
    if text:sub(pos, pos) ~= ":" then parse_error(text, pos, "expected ':'") end
    pos = skip_ws(text, pos + 1)
    v, pos = parse_value(text, pos)
    json.set(obj, k, v)
    pos = skip_ws(text, pos)
    local c = text:sub(pos, pos)
    if c == "," then
      pos = skip_ws(text, pos + 1)
    elseif c == "}" then
      return obj, pos + 1
    else
      parse_error(text, pos, "expected ',' or '}'")
    end
  end
end

parse_value = function(text, pos)
  pos = skip_ws(text, pos)
  local c = text:sub(pos, pos)
  if c == "{" then return parse_object(text, pos) end
  if c == "[" then return parse_array(text, pos) end
  if c == '"' then return parse_string(text, pos) end
  if c == "t" and text:sub(pos, pos + 3) == "true" then return true, pos + 4 end
  if c == "f" and text:sub(pos, pos + 4) == "false" then return false, pos + 5 end
  if c == "n" and text:sub(pos, pos + 3) == "null" then return json.null, pos + 4 end
  if c == "-" or c:match("%d") then return parse_number(text, pos) end
  parse_error(text, pos, "unexpected character " .. string.format("%q", c))
end

-- Decode a JSON document into the shape-preserving representation above.
function json.decode(text)
  local value, pos = parse_value(text, 1)
  pos = skip_ws(text, pos)
  if pos <= #text then parse_error(text, pos, "trailing garbage") end
  return value
end

return json

File: bindings/csharp/Causalontology/Json.cs
// A lossless JSON layer for the Causalontology C# binding.
//
// JSON values are represented as plain CLR values:
//   null, bool, long (integer literals), double (decimal literals),
//   string, List<object?> (arrays), and JsonMap (objects).
//
// The integer-versus-decimal source distinction (1 versus 1.0) survives
// to the canonicalizer: a literal with no '.', 'e', or 'E' becomes a
// long; anything else becomes a double. JsonMap keeps an explicit key
// insertion-order list - we deliberately do not rely on Dictionary's
// de-facto ordering.

using System.Globalization;
using System.Text;

namespace Causalontology;

/// <summary>A JSON object preserving key insertion order.</summary>
public sealed class JsonMap : IEnumerable<KeyValuePair<string, object?>>
{
    private readonly List<string> _order = new();
    private readonly Dictionary<string, object?> _values = new();

    public int Count => _order.Count;

    /// <summary>The keys in insertion order.</summary>
    public IReadOnlyList<string> Keys => _order;

    public object? this[string key]
    {
        get => _values[key];
        set
        {
            if (!_values.ContainsKey(key))
                _order.Add(key);
            _values[key] = value;
        }
    }

    /// <summary>Collection-initializer support: new JsonMap { { "k", v } }.</summary>
    public void Add(string key, object? value) => this[key] = value;

    public bool ContainsKey(string key) => _values.ContainsKey(key);

    public bool TryGetValue(string key, out object? value)
        => _values.TryGetValue(key, out value);

    /// <summary>Python dict.get(): the value, or null when absent.</summary>
    public object? Get(string key)
        => _values.TryGetValue(key, out var v) ? v : null;

    /// <summary>Python dict.get(key, default) for string values.</summary>
    public string? GetString(string key) => Get(key) as string;

    public bool Remove(string key)
    {
        if (!_values.Remove(key))
            return false;
        _order.Remove(key);
        return true;
    }

    /// <summary>Python dict.setdefault().</summary>
    public void SetDefault(string key, object? value)
    {
        if (!_values.ContainsKey(key))
            this[key] = value;
    }

    /// <summary>A shallow copy (Python dict(record)).</summary>
    public JsonMap Copy()
    {
        var copy = new JsonMap();
        foreach (var key in _order)
            copy[key] = _values[key];
        return copy;
    }

    public IEnumerator<KeyValuePair<string, object?>> GetEnumerator()
    {
        foreach (var key in _order)
            yield return new KeyValuePair<string, object?>(key, _values[key]);
    }

    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        => GetEnumerator();
}

/// <summary>A small lossless recursive-descent JSON parser.</summary>
public static class Json
{
    /// <summary>Parse a JSON document; integers stay long, decimals stay double.</summary>
    public static object? Parse(string text)
    {
        var pos = 0;
        var value = ParseValue(text, ref pos);
        SkipWhitespace(text, ref pos);
        if (pos != text.Length)
            throw new FormatException($"trailing data at position {pos}");
        return value;
    }

    /// <summary>Parse a JSON file.</summary>
    public static object? ParseFile(string path) => Parse(File.ReadAllText(path));

    /// <summary>Deep JSON equality; longs and doubles compare numerically (Python 1 == 1.0).</summary>
    public static bool DeepEquals(object? a, object? b)
    {
        if (a is null || b is null)
            return a is null && b is null;
        if (a is bool ba)
            return b is bool bb && ba == bb;
        if (IsNumber(a) && IsNumber(b))
            return ToDouble(a) == ToDouble(b);
        if (a is string sa)
            return b is string sb && sa == sb;
        if (a is List<object?> la)
        {
            if (b is not List<object?> lb || la.Count != lb.Count)
                return false;
            for (var i = 0; i < la.Count; i++)
                if (!DeepEquals(la[i], lb[i]))
                    return false;
            return true;
        }
        if (a is JsonMap ma)
        {
            if (b is not JsonMap mb || ma.Count != mb.Count)
                return false;
            foreach (var (key, value) in ma)
            {
                if (!mb.TryGetValue(key, out var other) || !DeepEquals(value, other))
                    return false;
            }
            return true;
        }
        return a.Equals(b);
    }

    /// <summary>True when the value is a JSON number (long, double, or int).</summary>
    public static bool IsNumber(object? value) => value is long or double or int;

    /// <summary>A JSON number as a double, for comparisons.</summary>
    public static double ToDouble(object? value) => value switch
    {
        long l => l,
        int i => i,
        double d => d,
        _ => throw new InvalidCastException($"not a number: {value}"),
    };

    private static void SkipWhitespace(string text, ref int pos)
    {
        while (pos < text.Length && text[pos] is ' ' or '\t' or '\n' or '\r')
            pos++;
    }

    private static object? ParseValue(string text, ref int pos)
    {
        SkipWhitespace(text, ref pos);
        if (pos >= text.Length)
            throw new FormatException("unexpected end of JSON");
        var ch = text[pos];
        switch (ch)
        {
            case '{': return ParseObject(text, ref pos);
            case '[': return ParseArray(text, ref pos);
            case '"': return ParseString(text, ref pos);
            case 't': Expect(text, ref pos, "true"); return true;
            case 'f': Expect(text, ref pos, "false"); return false;
            case 'n': Expect(text, ref pos, "null"); return null;
            default: return ParseNumber(text, ref pos);
        }
    }

    private static void Expect(string text, ref int pos, string literal)
    {
        if (pos + literal.Length > text.Length
            || text.Substring(pos, literal.Length) != literal)
            throw new FormatException($"invalid literal at position {pos}");
        pos += literal.Length;
    }

    private static JsonMap ParseObject(string text, ref int pos)
    {
        pos++; // consume '{'
        var map = new JsonMap();
        SkipWhitespace(text, ref pos);
        if (pos < text.Length && text[pos] == '}')
        {
            pos++;
            return map;
        }
        while (true)
        {
            SkipWhitespace(text, ref pos);
            var key = ParseString(text, ref pos);
            SkipWhitespace(text, ref pos);
            if (pos >= text.Length || text[pos] != ':')
                throw new FormatException($"expected ':' at position {pos}");
            pos++;
            map[key] = ParseValue(text, ref pos);
            SkipWhitespace(text, ref pos);
            if (pos >= text.Length)
                throw new FormatException("unterminated object");
            if (text[pos] == ',') { pos++; continue; }
            if (text[pos] == '}') { pos++; return map; }
            throw new FormatException($"expected ',' or '}}' at position {pos}");
        }
    }

    private static List<object?> ParseArray(string text, ref int pos)
    {
        pos++; // consume '['
        var list = new List<object?>();
        SkipWhitespace(text, ref pos);
        if (pos < text.Length && text[pos] == ']')
        {
            pos++;
            return list;
        }
        while (true)
        {
            list.Add(ParseValue(text, ref pos));
            SkipWhitespace(text, ref pos);
            if (pos >= text.Length)
                throw new FormatException("unterminated array");
            if (text[pos] == ',') { pos++; continue; }
            if (text[pos] == ']') { pos++; return list; }
            throw new FormatException($"expected ',' or ']' at position {pos}");
        }
    }

    private static string ParseString(string text, ref int pos)
    {
        if (text[pos] != '"')
            throw new FormatException($"expected string at position {pos}");
        pos++;
        var sb = new StringBuilder();
        while (true)
        {
            if (pos >= text.Length)
                throw new FormatException("unterminated string");
            var ch = text[pos++];
            if (ch == '"')
                return sb.ToString();
            if (ch != '\\')
            {
                sb.Append(ch);
                continue;
            }
            if (pos >= text.Length)
                throw new FormatException("unterminated escape");
            var esc = text[pos++];
            switch (esc)
            {
                case '"': sb.Append('"'); break;
                case '\\': sb.Append('\\'); break;
                case '/': sb.Append('/'); break;
                case 'b': sb.Append('\b'); break;
                case 'f': sb.Append('\f'); break;
                case 'n': sb.Append('\n'); break;
                case 'r': sb.Append('\r'); break;
                case 't': sb.Append('\t'); break;
                case 'u':
                    if (pos + 4 > text.Length)
                        throw new FormatException("truncated \\u escape");
                    sb.Append((char)ushort.Parse(
                        text.Substring(pos, 4), NumberStyles.HexNumber,
                        CultureInfo.InvariantCulture));
                    pos += 4;
                    break;
                default:
                    throw new FormatException($"invalid escape '\\{esc}'");
            }
        }
    }

    private static object ParseNumber(string text, ref int pos)
    {
        var start = pos;
        if (pos < text.Length && text[pos] == '-')
            pos++;
        while (pos < text.Length
               && (char.IsAsciiDigit(text[pos])
                   || text[pos] is '.' or 'e' or 'E' or '+' or '-'))
            pos++;
        var literal = text[start..pos];
        if (literal.Length == 0)
            throw new FormatException($"invalid number at position {start}");
        // the literal decides: no '.', 'e', or 'E' means an integer (long)
        if (!literal.Contains('.') && !literal.Contains('e') && !literal.Contains('E'))
            return long.Parse(literal, CultureInfo.InvariantCulture);
        return double.Parse(literal, CultureInfo.InvariantCulture);
    }
}
